### DISCUSSION TITLE
VectorLayer elements not redrawn on panning the map

### DISCUSSION THREAD

**Discussion by moving-bits** [mapsforge/vtm#998] [Q&A] [answered]
I'm struggling with a strange effect on a VTM `VectorLayer`:

- draw three paths (`LineDrawable`) in different locations (eg: near Berlin, Munich, and Wilhelmshaven)
- if map is zoomed out enough you can see, that all three elements are visible
- focus map on a single element
- pan map to one of the other elements => it's no longer being displayed
- pan map to the first element => it's also no longer being displayed
- if I check `size()` of the layer's `mDrawable` property, it still has three elements

Panning the map is done programmatically by calling the following method:
```
public void setCenter(final Geopoint geopoint) {
    final MapPosition pos = mMap.getMapPosition();
    pos.setPosition(geopoint.getLatitude(), geopoint.getLongitude());
    mMap.setMapPosition(pos);
}
```

Even calling `update()` on the affected layer does not help.

How can this happen? / What else can I check for?

**Comment by devemux86** [comment]
Can you share a [sample Activity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/PathOverlayActivity.java) to reproduce the report?

> calling update() on the affected layer

:+1: 

> Panning the map is done programmatically

Does it work when scroll the map with hand?

What happens if call `map.render()` or `map.updateMap()` or `map.clearMap()`?

**Reply by moving-bits** [reply]
Thanks for your very fast reply, very welcomed!

It seems to behave a bit different each time, probably depending on zoom level I start with when I open the map. (Map remembers the last zoom level.)

I've tried adding `map.render()`, `map.updateMap(true)` and `map.clearMap()` in `setCenter()`, but that made no difference.
Also, calling `update()` on the affected layer in `setCenter()` (instead of only in the place where it has been added to the layer the first time) did not resolve the problem.

Panning manually also made no difference,

One guess is, that only those objects are being drawn on panning, if they have been in the visible part of the map already at the time of initially adding it to the map.

---

I'll see if I can break this down into one of the sample apps. Not sure if I'll succeed, though, as we have a complex setting here with different tile providers, > 10 layers  etc., and an additional abstraction layer (to support a non-VTM map source as well), so I'll always have to combine the recipes of several of the sample apps.

Displaying the lines (tracks) is based on `LineDrawable` on a `VectorLayer`, as using a `PathLayer` gave me unwanted connections between the separate elements (see discussion in https://github.com/mapsforge/vtm/discussions/970).

My approach on adding / removing a track is like this:
- clear the layer (`trackLayer.clear()`, which does an `mDrawables.clear();`)
- add each segment to the layer: (`final LineDrawable path = new LineDrawable(segment, style); trackLayer.add(path);`)
- let the layer update (`trackLayer.update();`)

`segment` is an `ArrayList` of `GeoPoint`, `style` a `Style` with `strokeWidth` and `strokeColor`.

**Comment by devemux86** [comment]
@moving-bits thanks for using VTM in c:geo, so it has more users and we can improve it better!

I'm trying to remember if seeing such a case.
Except the needed `update()` on the layer after its creation (see the [VectorLayerActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/VectorLayerActivity.java)).

Testing this sample, we can see that the layers disappear when they start to go off the screen.

**Reply by moving-bits** [reply]
> Testing this sample, we can see that the layers disappear when they start to go off the screen.

Looks strange, indeed. Most of the time, the colored dots are available even when scrolling in from off-screen, but sometimes they are not.

Anything I can do about it in c:geo? In my test case the layer still holds the elements (mDrawables.size() == 3), but even calling `update()` on the layer does not bring it on screen. Anything I can check (eg a specific state field of an item)?

**Comment by devemux86** [comment]
> Looks strange, indeed.

It is necessary for performance.
Elements outside the screen should not be rendered (in all engines).

I do not remember if Mapsforge has a similar mechanism.

> Anything I can do about it in c:geo?

Sorry that I do not have quick answer...
A possible sample always help to make quick tests in code.

The only thing I can remember from Cruiser:
Sometimes rarely, when Cruiser starts, the restored route line is not rendered (only its waypoints appear).
If I scroll the map, the route line appears.

**Reply by moving-bits** [reply]
> > Looks strange, indeed.
> 
> It is necessary for performance. Elements outside the screen should not be rendered (in all engines).

I understand that, but shouldn't it be rendered (automatically) as soon as the formerly off-screen part enters the visible screen?

**Reply by devemux86** [reply]
> I understand that, but shouldn't it be rendered (automatically) as soon as the formerly off-screen part enters the visible screen?

I see this working with the [VectorLayerActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/VectorLayerActivity.java).
Circles disappear when exit the screen and appear again when enter the screen.

Do you see something different?

**Reply by moving-bits** [reply]
Yes, but it's not really the same as the problem described above:

When I move the map into one horizontal direction and get closer to the date line, the elements behind the date line are still missing. As soon as I cross the date line, the elements before it disappear and the elements behind it appear (so half the screen is empty shortly before and behind the date line).

**Reply by devemux86** [reply]
Ah, you had not mentioned the date line before. :slightly_smiling_face: 
Is the date line the issue of your original report?

---

I can reproduce the VectorLayerActivity on date line.
Probably the overlays are rendered on the map considered "active" each time (one of the two).
Could be related to #89.

---

Mapsforge does not have 2 maps at the date line: https://github.com/mapsforge/mapsforge/discussions/1361

**Reply by moving-bits** [reply]
> Ah, you had not mentioned the date line before. 🙂 Is the date line the issue of your original report?

No, all three elements of my original report are located within Germany, just a few hundred km away from each other, so that on a "normal" zoom level you won't see all three of them at the same time.

But the sample app does not really much more than adding a few elements with `layer.add()` - same as I do. One difference is, though, that the new `vectorLayer` is added to `mMap.layers()` after having been added the items, whereas in c:geo I do create all layers at opening of the map, and later on add/remove the elements as needed. But that should not make a difference, should it?

**Reply by moving-bits** [reply]
My use-case is:
- Let the app add all tracks to the map (in my example: three tracks, all within Germany)
- We have a function to "focus" on a track, meaning: center the map on the center of the track, keeping current zoom level, without animation
- When I start map pretty much zoomed out, I can see all three elements
- When I zoom in to one of them, the track is being displayed
- If I focus on a different track, the map moves to the correct location, but no track is being displayed. If I zoom out much more, no track is being displayed anymore. Some debug code revealed, that the trackLayer still holds three elements, they are just no longer being displayed - whereas elements from another layer (`ItemizedLayer` in this case) are still being displayed.

**Reply by devemux86** [reply]
Thanks for the details.

We need a small sample for the process (2 path overlays and move map with code).
Could use the [PathOverlayActivity](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/PathOverlayActivity.java) as an example (with `ANIMATION=false`).

So we can try to reproduce it with VTM code without any app influence.

**Reply by moving-bits** [reply]
I've created a sample based on `MapsforgeActivity` by adding the code listed below at the appropriate places, but have NOT been able to reproduce the behavior with that minimal case. So it's probably something in the specific environment of c:geo code, but how to find the bug? 🤔 

---

Adding to `MapsforgeActivity.java`:

somewhere at top of class:
```
    VectorLayer trackLayer;
    private final ArrayList<GeoPoint> track1 = new ArrayList<>(Arrays.asList(
            new GeoPoint(52.51, 13.3),
            new GeoPoint(52.52, 13.2),
            new GeoPoint(52.53, 13.4),
            new GeoPoint(52.54, 13.1)
    ));
    private final ArrayList<GeoPoint> track2 = new ArrayList<>(Arrays.asList(
            new GeoPoint(50.51, 13.3),
            new GeoPoint(50.52, 13.2),
            new GeoPoint(50.53, 13.4),
            new GeoPoint(50.54, 13.1)
    ));
    private final Style trackStyle = Style.builder()
            .strokeWidth(10)
            .strokeColor(0xaaff0000)
            .build();
    private static final int MENUID_ADD_LINES = 100;
    private static final int MENUID_FOCUS1 = 111;
    private static final int MENUID_FOCUS2 = 112;
```

add to `onCreateOptionsMenu`:
```
        menu.add(1, MENUID_ADD_LINES, 0, "Add lines");
        menu.add(1, MENUID_FOCUS1, 1, "Focus on track 1");
        menu.add(1, MENUID_FOCUS2, 2, "Focus on track 2");
```

add within case statement in `onOptionsItemSelected`:
```
            case MENUID_ADD_LINES: {
                trackLayer.add(new LineDrawable(track1, trackStyle));
                trackLayer.add(new LineDrawable(track2, trackStyle));
                trackLayer.update();
                return true;
            }
            case MENUID_FOCUS1: {
                setCenter(track1.get(0));
                return true;
            }
            case MENUID_FOCUS2: {
                setCenter(track2.get(0));
                return true;
            }
```

add in map init code in `onActivityResult`:
```
                trackLayer = new VectorLayer(mMap);
                mMap.layers().add(trackLayer);
```

add somewhere else within the class:
```
    public void setCenter(final GeoPoint geopoint) {
        final MapPosition pos = mMap.getMapPosition();
        pos.setPosition(geopoint.getLatitude(), geopoint.getLongitude());
        mMap.setMapPosition(pos);
    }
```

**Reply by moving-bits** [reply]
Using layer groups may be a solution to this issue, see https://github.com/cgeo/cgeo/pull/13806#issuecomment-1374460969
Will do some further testing and report back here.

**Comment by devemux86** [comment]
If you work with threads, perhaps some calls to VTM should be done with [Map.post(Runnable)](https://github.com/mapsforge/vtm/blob/master/vtm/src/org/oscim/map/Map.java#L261)?

On Android it uses the [View.post(Runnable)](https://developer.android.com/reference/android/view/View#post(java.lang.Runnable)): _"The runnable will be run on the user interface thread."_

**Reply by moving-bits** [reply]
Is there some documentation which methods may have the need to be called on UI thread?
(Or, a bit more in general: Is there some documentation on the core concepts of VTM, beside the source code?)

**Comment by devemux86** [comment]
> Is there some documentation on the core concepts of VTM, beside the source code?

All docs are linked in the [main page](https://github.com/mapsforge/vtm) and exist in the [docs](https://github.com/mapsforge/vtm/tree/master/docs) folder.

I am afraid for more details all of us usually need to look the code. :slightly_smiling_face:

**Comment by devemux86** [comment, marked as answer]
https://github.com/cgeo/cgeo/pull/13806#issuecomment-1374460969

**Reply by moving-bits** [reply]
And https://github.com/cgeo/cgeo/pull/13833, merged today. Let's see how it works out "in the wild" :-)
